### PR TITLE
add version string to metric-proxy binary

### PR DESCRIPTION
**Pull request by dtimm**
- add Makefile that creates usable tag
- use make in Dockerfile

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -20,18 +20,22 @@ import (
...
+var (
+	version = "dev-build"
+	requestDurations metricRegistry.Histogram
+)
+
 
 func main() {
 	loggr := log.New(os.Stderr, "", log.LstdFlags)
+	loggr.Printf("starting metric-proxy %s...\n", version)
+	defer loggr.Println("exiting metric-proxy...")
```
Comment (louis-brann) [inline main.go:32]: nice!

File: Dockerfile [hunk 2]
```diff
@@ -2,20 +2,15 @@ FROM golang:1.13 as builder
 
 COPY / /metric-proxy/
 
-RUN cd /metric-proxy && go build \
-    -a \
-    -installsuffix nocgo \
-    -o /bin/metric-proxy \
-    -mod=readonly \
-    .
+RUN cd /metric-proxy && make
```
Comment (louis-brann) [inline Dockerfile:5]: nice!

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# With thanks to:
+# https://dev.to/eugenebabichenko/generating-pretty-version-strings-including-nightly-with-git-and-makefiles-48p3
+TAG_COMMIT := $(shell git rev-list --abbrev-commit --tags --max-count=1)
+TAG := $(shell git describe --abbrev=0 --tags ${TAG_COMMIT} 2>/dev/null || true)
+COMMIT := $(shell git rev-parse --short HEAD)
+DATE := $(shell git log -1 --format=%cd --date=format:"%Y%m%d")
+VERSION := $(TAG:v%=%)
```
Comment (heycait) [inline Makefile:7]: Is this checking for a "v" in the tag?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	echo "Building windows binary tag $(VERSION) in bin/metric-proxy-windows..."
+	GOOS=windows GOARCH=amd64 go build -a -installsuffix nocgo -o bin/metric-proxy-windows -ldflags "-X main.version=$(VERSION)" -mod=readonly .
+
+build-darwin:
+	echo "Building darwin binary tag $(VERSION) in bin/metric-proxy-darwin..."
+	GOOS=darwin OARCH=amd64 go build -a -installsuffix nocgo -o bin/metric-proxy-darwin -ldflags "-X main.version=$(VERSION)" -mod=readonly .
+
+test:
+	go test ./... -race -count=1
+
```
Comment (heycait) [inline Makefile:37]: Does this mean we can run `make test` to get go test to run?

Comment (dtimm) [reply Makefile:37]: Yeah, exactly.

File: Makefile (hunk 3 above)
Comment (dtimm) [reply Makefile:7]: It strips a leading v from version tag if there is one.

File: Makefile (hunk 4 above)
Comment (heycait) [reply Makefile:37]: cool cool

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+build-all: build-linux build-windows build-darwin
+
+build-linux:
+	echo "Building linux binary tag $(VERSION) in bin/metric-proxy-linux..."
+	GOOS=linux GOARCH=amd64 go build -a -installsuffix nocgo -o bin/metric-proxy-linux -ldflags "-X main.version=$(VERSION)" -mod=readonly .
+
+build-windows:
+	echo "Building windows binary tag $(VERSION) in bin/metric-proxy-windows..."
+	GOOS=windows GOARCH=amd64 go build -a -installsuffix nocgo -o bin/metric-proxy-windows -ldflags "-X main.version=$(VERSION)" -mod=readonly .
```
[the remark this answers is not in the document]
Comment (louis-brann) [reply Makefile:29]: there's probably some fancy stuff we could do like [prometheus/statsd_exporter](https://github.com/prometheus/statsd_exporter/blob/master/Makefile.common#L236), but this is good for readability

**Pull request merged**